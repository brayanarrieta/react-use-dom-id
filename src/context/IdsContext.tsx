import React from 'react';
import { DEFAULT_ID_SEPARATOR, DEFAULT_INITIAL_ID_NAMESPACE_COUNTER, DEFAULT_NAMESPACE } from '../constants';

let counterNamespaces : {[key: string]: number} = {};

const IdsContext = React.createContext({
  generateId(namespacePrefix?: string) {
    const namespace = namespacePrefix || DEFAULT_NAMESPACE;

    if (!counterNamespaces[namespace]) {
      counterNamespaces[namespace] = DEFAULT_INITIAL_ID_NAMESPACE_COUNTER;
    }
    counterNamespaces[namespace] += 1;

    return `${namespace}${DEFAULT_ID_SEPARATOR}${counterNamespaces[namespace]}`;
  },
});

export const resetIdsContext = () => {
  counterNamespaces = {};
};

export default IdsContext;
