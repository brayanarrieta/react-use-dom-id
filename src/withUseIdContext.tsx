import React from 'react';
import useIdContext from './hooks/useIdContext';

const withUseIdContext = (
  namespacePrefixContext?: string,
) => (Component: React.ComponentType<any>) => {
  // eslint-disable-next-line react/function-component-definition
  const WrappedComponent = (props: any) => {
    const { generateId: generateIdContext } = useIdContext();

    const generateId = (
      namespacePrefix?: string,
    ) => generateIdContext(namespacePrefix || namespacePrefixContext);

    return (
      <Component
        // eslint-disable-next-line react/jsx-props-no-spreading
        {...props}
        useIdContext={{
          generateId,
        }}
      />
    );
  };

  return WrappedComponent;
};

export default withUseIdContext;
