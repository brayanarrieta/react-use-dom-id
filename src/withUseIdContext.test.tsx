import React from 'react';
import { render } from '@testing-library/react';
import withUseIdContext from './withUseIdContext';
import { resetIdsContext } from './context/IdsContext';

const generateMockComponent = (
  namespace?: string,
// eslint-disable-next-line react/function-component-definition
) => (props: {useIdContext: { generateId: any }}) => {
  const { useIdContext } = props;
  const { generateId } = useIdContext;
  return <div id={generateId(namespace)} />;
};

describe('withUseId', () => {
  afterEach(() => {
    resetIdsContext();
  });

  it('when not pass a namespace prefix should generate a unique id with the default namespace and pass as id prop', () => {
    const MockComponent = generateMockComponent();
    const WrappedComponent = withUseIdContext()(MockComponent);
    const { container } = render(<WrappedComponent />);
    expect(container).toMatchSnapshot();
  });

  it('when not pass namespaces should generate unique ids', () => {
    const FirstMockComponent = generateMockComponent();
    const FirstWrappedComponent = withUseIdContext()(FirstMockComponent);
    const { container: firstContainer } = render(<FirstWrappedComponent />);
    expect(firstContainer).toMatchSnapshot();

    const SecondMockComponent = generateMockComponent();
    const SecondWrappedComponent = withUseIdContext()(SecondMockComponent);
    const { container: secondContainer } = render(<SecondWrappedComponent />);
    expect(secondContainer).toMatchSnapshot();
  });

  describe('when pass a namespace prefix context', () => {
    it('it should be part of the unique id', () => {
      const MockComponent = generateMockComponent();
      const WrappedComponent = withUseIdContext('test-namespace')(MockComponent);
      const { container } = render(<WrappedComponent />);
      expect(container).toMatchSnapshot();
    });

    it('and override with the namespace prefix param it the new one should be part of the unique id', () => {
      const MockComponent = generateMockComponent('new-test-namespace');
      const WrappedComponent = withUseIdContext('test-namespace')(MockComponent);
      const { container } = render(<WrappedComponent />);
      expect(container).toMatchSnapshot();
    });
  });
});
