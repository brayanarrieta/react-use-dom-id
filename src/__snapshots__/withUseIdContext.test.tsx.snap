// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withUseId when not pass a namespace prefix should generate a unique id with the default namespace and pass as id prop 1`] = `
<div>
  <div
    id="app-1"
  />
</div>
`;

exports[`withUseId when not pass namespaces should generate unique ids 1`] = `
<div>
  <div
    id="app-1"
  />
</div>
`;

exports[`withUseId when not pass namespaces should generate unique ids 2`] = `
<div>
  <div
    id="app-2"
  />
</div>
`;

exports[`withUseId when pass a namespace prefix context and override with the namespace prefix param it the new one should be part of the unique id 1`] = `
<div>
  <div
    id="new-test-namespace-1"
  />
</div>
`;

exports[`withUseId when pass a namespace prefix context it should be part of the unique id 1`] = `
<div>
  <div
    id="test-namespace-1"
  />
</div>
`;
